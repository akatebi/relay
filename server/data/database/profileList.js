import myFetch from './myFetch';
import { serverPath, fetchOptions } from './constant';

export const getProfileList = (token, entity) => {
  const url = `${serverPath}${entity}`;
  const options = fetchOptions(token);
  return myFetch(url, options);
};
