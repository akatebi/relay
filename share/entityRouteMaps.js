import {
  entityTypeToRevisionRoute,
  entityTypeToFamilyRoute,
  entityTypeToSimpleRoute,
  entityTypeToTreeNode,
  entityLabelToFamilyRoute,
  entityListLabels,
  entityTypeToLabel,
  entityTypeToPluralLabel,
  revisionableList,
  routeToEntityType,
} from './entityTypes';

// const debug = require('debug')('app:share:entityRouteMaps');

// this returns route that returns the ResultSetVM
export const entityRevRouteMap = (entityType) => {
  const route = entityTypeToRevisionRoute[entityType];
  // debug('entityRevRouteMap', entityType, route);
  if (!route) {
    const msg = `entityRevRouteMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityFamilyRouteMap = (entityType) => {
  const route = entityTypeToFamilyRoute[entityType];
  // debug('entityFamilyRouteMap', entityType, route);
  if (!route) {
    const msg = `entityFamilyRouteMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

// this returns route that returns an EntityVM list
export const entityRevSimpleMap = (entityType) => {
  const route = entityTypeToSimpleRoute[entityType];
  if (!route) {
    const msg = `entityRevSimpleMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityTreeNodeRouteMap = (entityType) => {
  // console.log('entityTreeNodeRouteMap', entityType);
  const route = entityTypeToTreeNode[entityType];
  if (!route) {
    const msg = `entityTreeNodeRouteMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityCreationFamilyRouteMap = (entityLabel) => {
  const route = entityLabelToFamilyRoute[entityLabel];
  if (!route) {
    const msg = `entityLabelFamily(${entityLabel}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityRouteToTypeMap = (pathname) => {
  const entityRevRoute = pathname.split('/')[1];
  const entityType = routeToEntityType[entityRevRoute];
  if (!entityType) {
    const msg = `routeToEntityType(${entityRevRoute}) => undefined`;
    console.error(msg);
  }
  return entityType;
};

export const entityTypeToTreeNodeTypeMap = (entityType) => {
  const treeNodeType = entityTypeToTreeNode[entityType];
  if (!treeNodeType) {
    const msg = `routeToEntityType(${entityType}) => undefined`;
    console.error(msg);
  }
  return treeNodeType;
};

export const getEntityLabelByType = (entityType) => {
  const label = entityTypeToLabel[entityType];
  if (!label) {
    const msg = `entityListLabels has no "${entityType}"`;
    console.error(msg);
  }
  return label;
};

export const getEntityListLabel = () => {
  const entity = location.pathname.slice(1).split('/')[0];
  const key = Object.keys(entityListLabels).find(label =>
    new RegExp(label).test(entity));
  return key ? entityListLabels[key] : 'Unknown Entity';
};

export const getEntityPluralLabelByType = (entityType) => {
  const label = entityTypeToPluralLabel[entityType];
  if (!label) {
    const msg = `getEntityPluralLabelByType(${entityType}) => undefined`;
    console.error(msg);
  }
  return label;
};

export const getEntityPluralLabelByRoute = (pathname) => {
  const entityType = entityRouteToTypeMap(pathname);
  return getEntityPluralLabelByType(entityType);
};

export const isRevisionable = entity => revisionableList.includes(entity);

export const getEntityUrl = entityType =>
  (/LifeCycleConfig|User/.test(entityType) ?
    entityFamilyRouteMap(entityType) :
    entityRevRouteMap(entityType));
