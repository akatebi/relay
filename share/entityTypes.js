
export const EntityTypes = [
  'Action',
  'ApplicationRole',
  'ApprovalCycleConfig',
  'AssociationConfig',
  'CategoryHierarchy',
  'CharacteristicConfig',
  'ChoiceList',
  'DictionaryEntry',
  'Document',
  'DocumentClass',
  'DocumentType',
  'Layout',
  'LifeCycleConfig',
  'LineItemConfig',
  'NavigationHierarchy',
  'NavigationNode',
  'OrganizationHierarchy',
  'Organization',
  'SectionConfig',
  'User',
];

// EntityType ===>> Revision Route
export const entityTypeToRevisionRoute = {
  Action: 'actions/revisions',
  ApplicationRole: 'applicationroles/revisions',
  ApprovalCycleConfig: 'approvalcycleconfigs/revisions',
  AssociationConfig: 'associationconfigs/revisions',
  CategoryHierarchy: 'categoryhierarchies/revisions',
  CharacteristicConfig: 'characteristicconfigs/revisions',
  ChoiceList: 'choicelists/revisions',
  DictionaryEntry: 'dictionaryentries/revisions',
  Document: 'documents/revisions',
  DocumentClass: 'documentclasses/revisions',
  DocumentType: 'documenttypes/revisions',
  Layout: 'templatelayouts/revisions',
  LifeCycleAction: 'lifecycles/actions',
  LifeCycleConfig: 'lifecycleconfigs',
  LineItemConfig: 'lineitemconfigs/revisions',
  NavigationHierarchy: 'navigationhierarchies/revisions',
  NavigationNode: 'navigationnodes/revisions',
  OrganizationHierarchy: 'organizationhierarchies/revisions',
  Organization: 'organizations/revisions',
  SectionConfig: 'sectionconfigs/revisions',
  User: 'users',
};

// EntityType ===>> Family Route
export const entityTypeToFamilyRoute = {
  Action: 'actions',
  ApplicationRole: 'applicationroles',
  ApprovalCycleConfig: 'approvalcycleconfigs',
  AssociationConfig: 'associationconfigs',
  CategoryHierarchy: 'categoryhierarchies',
  CharacteristicConfig: 'characteristicconfigs',
  ChoiceList: 'choicelists',
  DictionaryEntry: 'dictionaryentries',
  Document: 'documents',
  DocumentClass: 'documentclasses',
  DocumentType: 'documenttypes',
  Layout: 'templatelayouts',
  LifeCycleAction: 'lifecycles/actions',
  LifeCycleConfig: 'lifecycleconfigs',
  LineItemConfig: 'lineitemconfigs',
  NavigationHierarchy: 'navigationhierarchies',
  NavigationNode: 'navigationnodes',
  OrganizationHierarchy: 'organizationhierarchies',
  Organization: 'organizations',
  SectionConfig: 'sectionconfigs',
  User: 'users',
};

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

export const entityTypeToPluralLabel = {
  Action: 'Actions',
  ApplicationRole: 'Application Roles',
  ApprovalCycleConfig: 'Approval Cycle Configurations',
  AssociationConfig: 'Association Configurations',
  CharacteristicConfig: 'Characteristic Configurations',
  CategoryHierarchy: 'Category Hierarchies',
  ChoiceList: 'Choice Lists',
  Culture: 'Cultures',
  DictionaryEntry: 'Dictionary Entries',
  Document: 'Documents',
  DocumentClass: 'Document Classes',
  DocumentType: 'Document Types',
  Layout: 'Template Layouts',
  LifeCycleConfig: 'Life Cycle Configurations',
  LineItemConfig: 'Line Item Configurations',
  NavigationHierarchy: 'NavigationHierarchies',
  NavigationNode: 'NavigationNodes',
  OrganizationHierarchy: 'Organization Hierarchies',
  Organization: 'Organizations',
  SectionConfig: 'Section Configurations',
  SectionLineItemConfig: 'Section Line Item Configurations',
  User: 'User Profiles',
};

export const entityListLabels = {
  Action: 'Action',
  ApplicationRole: 'Application Role',
  ApprovalCycleConfig: 'Approval Cycle Configuration',
  Assignment: 'Assignment',
  AssociationConfig: 'Association Configuration',
  CategoryHierarchy: 'Category Hierarchy',
  CharacteristicConfig: 'Characteristic Configuration',
  ChoiceList: 'Choice List',
  DictionaryEntry: 'Dictionary Entry',
  Culture: 'Culture',
  Document: 'Document',
  DocumentClass: 'Document Class',
  DocumentType: 'Document Type',
  LifeCycleConfig: 'Life Cycle Configuration',
  LineItemConfig: 'Line Item Configuration',
  NavigationNode: 'Navigation Node',
  NavigationHierarchy: 'Navigation Hierarchy',
  Organization: 'Organization',
  OrganizationHierarchy: 'Organization Hierarchy',
  SectionConfig: 'Section Configuration',
  TemplateLayout: 'Template Layout',
  User: 'User Profile',
};

// EntityType ===>> Simple Route
export const entityTypeToSimpleRoute = {
  Action: 'actions/simple',
  ApplicationRole: 'applicationroles/simple',
  ApprovalCycleConfig: 'approvalcycleconfigs/simple',
  AssociationConfig: 'associationconfigs/simple',
  CategoryHierarchy: 'categoryhierarchies/simple',
  CharacteristicConfig: 'characteristicconfigs/simple',
  ChoiceList: 'choicelists/simple',
  DictionaryEntry: 'dictionaryentries/simple',
  Document: 'documents/simple',
  DocumentClass: 'documentclasses/simple',
  DocumentType: 'documenttypes/simple',
  Layout: 'templatelayouts/simple',
  LifeCycleAction: 'lifecycles/actions/simple',
  LifeCycleConfig: 'lifecycles/simple',
  LineItemConfig: 'lineitemconfigs/simple',
  NavigationHierarchy: 'navigationhierarchies/simple',
  NavigationNode: 'navigationnodes/simple',
  OrganizationHierarchy: 'organizationhierarchies/simple',
  Organization: 'organizations/simple',
  SectionConfig: 'sectionconfigs/simple',
  User: 'users/simple',
};

// EntityLabel ===>> Family Route
export const entityLabelToFamilyRoute = {
  Action: 'actions',
  'Application Role': 'applicationroles',
  'Approval Cycle Config': 'approvalcycleconfigs',
  'Association Config': 'associationconfigs',
  'Category Hierarchy': 'categoryhierarchies',
  'Characteristic Config': 'characteristicconfigs',
  'Choice List': 'choicelists',
  'Dictionary Entry': 'dictionaryentries',
  Document: 'documents',
  'Document Class': 'documentclasses',
  Layout: 'templatelayouts',
  'Lifecycle Config': 'lifecycleconfigs',
  'Line Item Config': 'lineitemconfigs',
  'Navigation Hierarchy': 'navigationhierarchies',
  'Navigation Node': 'navigationnodes',
  Organization: 'organizations',
  'Section Config': 'sectionconfigs',
  User: 'users',
};

// Route ===>> Entity Type
export const routeToEntityType = {
  actions: 'Action',
  applicationroles: 'ApplicationRole',
  approvalcycleconfigs: 'ApprovalCycleConfig',
  associationconfigs: 'AssociationConfig',
  categoryhierarchies: 'CategoryHierarchy',
  characteristicconfigs: 'CharacteristicConfig',
  choicelists: 'ChoiceList',
  dictionaryentries: 'DictionaryEntry',
  documents: 'Document',
  documentclasses: 'DocumentClass',
  documenttypes: 'DocumentType',
  lifecycleconfigs: 'LifeCycleConfig',
  lineitemconfigs: 'LineItemConfig',
  navigationhierarchies: 'NavigationHierarchy',
  navigationnodes: 'NavigationNode',
  organizationhierarchies: 'OrganizationHierarchy',
  organizations: 'Organization',
  sectionconfigs: 'SectionConfig',
  templatelayouts: 'TemplateLayout',
  users: 'User',
};

// Entity Type ===>> Tree Node Type
export const entityTypeToTreeNode = {
  CategoryHierarchy: 'dictionaryentries',
  ChoiceList: 'dictionaryentries',
  NavigationHierarchy: 'navigationnodes',
  OrganizationHierarchy: 'organizations',
};


export const entityTypeToLabel = {
  Action: 'Action',
  ApprovalCycleConfig: 'Approval Cycle Configuration',
  ApplicationRole: 'Application Role',
  Assignment: 'Assignment',
  AssociationConfig: 'Association Configuration',
  CategoryHierarchy: 'Category Hierarchy',
  CharacteristicConfig: 'Characteristic Configuration',
  ChoiceList: 'Choice List',
  Culture: 'Culture',
  DictionaryEntry: 'Dictionary Entry',
  Document: 'Document',
  DocumentClass: 'Document Class',
  DocumentType: 'Document Type',
  LifeCycleConfig: 'Life Cycle Configuration',
  LineItemConfig: 'Line Item Configuration',
  NavigationHierarchy: 'Navigation Hierarchy',
  NavigationNode: 'Navigation Node',
  OrganizationHierarchy: 'Organization Hierarchy',
  Organization: 'Organization',
  SectionConfig: 'Section Configuration',
  SectionLineItemConfig: 'Section Line Item Configuration',
  TemplateLayout: 'Template Layout',
  User: 'User Profile',
};

export const supportedCreationEntities = [
  'Action',
  'Application Role',
  'Approval Cycle Config',
  'Association Config',
  'Category Hierarchy',
  'Characteristic Config',
  'Choice List',
  'Dictionary Entry',
  'Document',
  'Document Class',
  'Layout',
  'Line Item Config',
  'Navigation Node',
  'Navigation Hierarchy',
  'Organization',
  'Section Config',
  'Sequence',
  'User',
];

export const propertyTypes = [
  'Boolean',
  'Category',
  // 'ControlNumber',
  'Date',
  'Decimal',
  'Entity',
  'EntityList',
  'Float',
  'Integer',
  'RoleList',
  'RichText',
  'Text',
];

export const revisionableList = [
  'actions',
  'applicationroles',
  'approvalcycleconfigs',
  'associationconfigs',
  'categoryhierarchies',
  'characteristicconfigs',
  'choicelists',
  'dictionaryentries',
  'documents',
  'documentclasses',
  'documenttypes',
  'lineitemconfigs',
  'navigationhierarchies',
  'navigationnodes',
  'organizations',
  'organizationhierarchies',
  'sectionconfigs',
  'templatelayouts',
];
