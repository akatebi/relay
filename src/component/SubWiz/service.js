
export const selectPropsMap = list =>
  list.map(item => ({
    value: item.id ? item.id.slice(0, 36) : item.value,
    label: item.label,
    cultureCode: item.cultureCode,
    entityType: item.entityType,
    familyId: item.familyId,
    id: item.id,
    isFamily: item.isFamily,
    operation: item.operation,
    reportingKey: item.reportingKey,
  }));

// used by AppSegments
const getSegmentList = (approvalCycle) => {
  const segments = approvalCycle.approvalSegmentConfigs;

  const result = segments.length ?
    segments.map(segment => ({
      id: segment.id,
      segmentType: segment.segmentType,
      segmentLabel: segment.label,
      waitForEverybody: segment.waitForEverybody,
      placeholder: 'select approvers...',
      approverSelectionMode: segment.selectAllUsers ? 'Select Users' : 'All Matching Users',
      approverSource: segment.approverSource,
      approvers: {
        eligible: [],
        selected: [],
      },
    })) :
    ({
      id: '',
      segmentType: 'none',
      segmentLabel: 'none',
      waitForEverybody: false,
      placeholder: 'none',
      approverSelectionMode: 'none',
      approverSource: 'none',
      approvers: {
        eligible: [{ value: 'none', label: 'none' }],
        selected: [],
      },
    });
  // console.log('>>> getSegmentList result', window.pretty(result));
  return result;
};

// used by AppSegments
export const approvalCycleMap = approvalCycle =>
  ({
    id: approvalCycle.id,
    entityType: approvalCycle.entityType,
    label: approvalCycle.label,
    processType: approvalCycle.processType,
    waitForEverybody: approvalCycle.waitForEverybody,
    segmentList: getSegmentList(approvalCycle),
  });

const getAssigneesNew = (selectedApprovers) => {
  // console.log('selectedApprovers', window.pretty(selectedApprovers));
  const list = selectedApprovers.map(approver =>
    approver.reduce((acc, { segmentId, id }) => {
      acc.segmentId = segmentId;
      const obj = { id: id.slice(0, 36), isFamily: true, entityType: 'ApplicationRole' };
      acc.assignees.push(obj);
      return acc;
    }, { assignees: [] }));
  // console.log('list', window.pretty(list));
  return list;
};

const getAppCycleConfig = selectedCycleId => ({
  isFamily: true,
  id: selectedCycleId,
  entityType: 'ApprovalCycleConfig',
});

// used by AppSegments.onSubmit
// ==============================
export const mapSubmitData = ({
  lifeCycle,
  processType,
  selectedCycleId,
  selectedApprovers,
}) => {
  let toState = null;
  if (processType === 'approval') {
    toState = lifeCycle.lifeCycleStateContents
      .find(stateContent => stateContent.state === 'Approved');
    delete toState.__id;
  }
  return {
    toState,
    directApproval: false,
    isFamily: true,
    segmentAssignees: getAssigneesNew(selectedApprovers),
    approvalCycleConfig: getAppCycleConfig(selectedCycleId),
  };
};
