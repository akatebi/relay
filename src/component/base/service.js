
// const debug = require('debug')('app:root:service');

export const getGUID = pathname =>
  pathname.split('/').find(item =>
    /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/.test(item));

export const allParams = ({ location, match, params = {}, name = '' }) => {
  const path = `${match.path}/${name}`.slice(1);
  const pathname = location.pathname.slice(1);
  const pathList = path.split('/');
  const pathnameList = pathname.split('/');
  // debug(name, 'pathList', pathList);
  // debug(name, 'pathnameList', pathnameList);
  const newParams = pathList.reduce((acc, prop, index) => {
    const value = pathnameList[index];
    // debug(name, 'prop', prop);
    if (prop[0] === ':') {
      acc[prop.slice(1)] = value;
    }
    return acc;
  }, {});
  // debug(name, 'newParams', window.pretty(newParams));
  const paramsAll = { ...params, ...newParams };
  // debug(name, 'params keys', Object.keys(paramsAll));
  return paramsAll;
};
