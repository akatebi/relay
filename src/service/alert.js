import React from 'react';
import qs from 'qs';
import Alert from 'react-s-alert';
import moment from 'moment';

// const debug = require('debug')('app:service:alert');

const options = {
  position: 'top-right',
  effect: 'slide',
  beep: false,
  timeout: 3000,
  offset: 0,
};

const timeString = (timeStr) => {
  const fromNow = `${moment(timeStr).fromNow()}`;
  const dateFmt = `${moment(timeStr).format('ll')} ${moment(timeStr).format('LT')}`;
  const dateOut = `${fromNow} on ${dateFmt}`;
  return dateOut;
};

const getMsg = (msg, eventTime) =>
  (eventTime ? `${msg} - ${timeString(eventTime)}` : msg);

export const successAlert = (msg, eventTime) => {
  setTimeout(() => {
    const message = getMsg(msg, eventTime);
    Alert.success(message, options);
  }, 1000);
};

export const infoAlert = (msg, eventTime) => {
  let message;
  if (!eventTime || eventTime === '0001-01-01T00:00:00Z') {
    message = msg;
  } else {
    message = getMsg(msg, eventTime);
  }
  Alert.info(message, options);
};

export const errorAlert = (msg, eventTime) => {
  const message = getMsg(msg, eventTime);
  if (/^https/.test(message)) {
    const { id } = qs.parse(message.split('_queries?')[1]);
    Alert.error(<a target="foo" href={message}>{`BUG: ${id}`}</a>, { ...options, timeout: 10000 });
  } else {
    Alert.error(message, { ...options, timeout: 10000 });
  }
};
