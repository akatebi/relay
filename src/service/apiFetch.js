import deepmerge from 'deepmerge';
import { serverPath, getToken } from '../constant/app';

const debug = require('debug')('app:service:apiFetch');

function apiFetch(apiUrl, extraOptions = {}) {
  const url = `${serverPath}/${apiUrl}`;
  const Authorization = `Bearer ${getToken()}`;
  const options = deepmerge({
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization,
    },
  }, extraOptions);
  debug(url);
  // debug(options);
  return fetch(url, options)
    .then((resp) => {
      if (!resp.ok) {
        const error = { error: `${resp.status}-${resp.statusText}` };
        throw error;
      }
      // debug('resp:', resp);
      return resp;
    })
    .catch((error) => {
      debug('API-FETCH Error:', error);
      throw error;
    });
}

export default apiFetch;

export const myFetch = (url, extraOptions = {}) => {
  const Authorization = `Bearer ${getToken()}`;
  const options = deepmerge({
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization,
    },
  }, extraOptions);
  // debug(url);
  // debug(options);
  return fetch(url, options);
};
