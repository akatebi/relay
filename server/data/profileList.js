import {
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
  GraphQLUnionType,
} from 'graphql';

import {
  getProfileList,
} from './database/profileList';

import { getToken } from './database/constant';

import { GraphQLEntityVMType } from './entityVM';

const GraphQLColumnHeaderType = new GraphQLObjectType({
  name: 'ListColumnHeader',
  fields: {
    index: { type: new GraphQLNonNull(GraphQLInt) },
    label: { type: new GraphQLNonNull(GraphQLString) },
    type: { type: new GraphQLNonNull(GraphQLString) },
  },
});

const ControlNumberColumnValueType = new GraphQLObjectType({
  name: 'ControlNumberColumnValue',
  fields: {
    valueControlNumber: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: ({ value: { controlNumberString } }) => controlNumberString,
    },
  },
});

const DateColumnValueType = new GraphQLObjectType({
  name: 'DateColumnValue',
  fields: {
    valueDate: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: ({ value }) => value,
    },
  },
});

const RoleListColumnValueType = new GraphQLObjectType({
  name: 'RoleListColumnValue',
  fields: {
    valueRoleList: {
      type: new GraphQLNonNull(new GraphQLList(GraphQLEntityVMType)),
      resolve: ({ value: { $values } }) => $values,
    },
  },
});

const TextColumnValueType = new GraphQLObjectType({
  name: 'TextColumnValue',
  fields: {
    valueText: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: ({ value }) => value,
    },
  },
});

export const GraphQLColumnValueUnionType = new GraphQLUnionType({
  name: 'ProfileTableColumnValueUnion',
  types: [
    ControlNumberColumnValueType,
    DateColumnValueType,
    RoleListColumnValueType,
    TextColumnValueType,
  ],
  resolveType: ({ propertyType }) => {
    switch (propertyType) {
      case 'ControlNumber':
        return ControlNumberColumnValueType;
      case 'Date':
        return DateColumnValueType;
      case 'RoleList':
        return RoleListColumnValueType;
      case 'Text':
        return TextColumnValueType;
      default:
        console.error(`${propertyType} is unknown`);
        return null;
    }
  },
});

const GraphQLColumnType = new GraphQLObjectType({
  name: 'ProfileTableColumn',
  fields: {
    index: { type: new GraphQLNonNull(GraphQLInt) },
    propertyType: { type: new GraphQLNonNull(GraphQLString) },
    value: {
      type: GraphQLColumnValueUnionType,
      resolve: ({ propertyType, value }) => ({ propertyType, value }),
    },
  },
});

const GraphQLRowType = new GraphQLObjectType({
  name: 'ProfileTableRow',
  fields: {
    rowKey: { type: new GraphQLNonNull(GraphQLEntityVMType) },
    columns: {
      type: new GraphQLNonNull(new GraphQLList(GraphQLColumnType)),
    },
  },
});

export const GraphQLProfileTableType = new GraphQLObjectType({
  name: 'ProfileTable',
  fields: {
    sortedColumnIndex: { type: new GraphQLNonNull(GraphQLInt) },
    sortOrder: { type: new GraphQLNonNull(GraphQLString) },
    columnHeaders: {
      type: new GraphQLNonNull(new GraphQLList(GraphQLColumnHeaderType)),
    },
    rows: {
      type: new GraphQLNonNull(new GraphQLList(GraphQLRowType)),
      resolve: ({ rows, columnHeaders }) => {
        rows.forEach(row => row.columns.forEach((col, i) => {
          col.propertyType = columnHeaders[i].type;
        }));
        return rows;
      },
    },
  },
});

const GraphQLProfileList = new GraphQLObjectType({
  name: 'ProfileList',
  fields: {
    tables: {
      type: new GraphQLNonNull(new GraphQLList(GraphQLProfileTableType)),
    },
  },
});

export const profileList = {
  type: GraphQLProfileList,
  args: {
    entity: { type: GraphQLString },
  },
  resolve: (...args) => {
    const token = getToken(args[2], args[0]);
    const { entity } = args[1];
    return getProfileList(token, entity);
  },
};
