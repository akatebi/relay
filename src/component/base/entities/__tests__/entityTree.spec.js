import clone from 'clone';
import { getToken, cleanLog, writeFile, graphql, fixGUIDS } from '../../../__tests__/service';
import { EntityTypes } from '../../../../../share/entityTypes';
import { getProfileList } from '../../../../../server/data/database/profileList';
import { entityRevRouteMap } from '../../../../../share/entityRouteMaps';
import query from './query/entityTree';

const debug = require('debug')('app:src:base:entities:entityTree');

const dirname = `${__dirname}/logs/entityTree`;
const writeTB = writeFile.bind(this, dirname);

////////////////////// M O D I F Y V A R I A B L E S ////////////////////////////
const docCount = 1;
const entityTypeRegExp = /OrganizationHierarchy/;
/////////////////////////////////////////////////////////////////////////////////

describe('entityTree', () => {

  beforeAll(getToken);
  beforeAll(cleanLog.bind(this, dirname));

  EntityTypes.filter(en =>
    entityTypeRegExp.test(en)).map(async (entityType) => {
    test(entityType, async () => {
      const { token } = global;
      const entity = `/${entityRevRouteMap(entityType)}/ids`;
      const profileList = await getProfileList(token, entity);
      expect(fixGUIDS(clone(profileList))).toMatchSnapshot();
      // debug('profileList', profileList);
      await writeTB(entityType, 'profileList', profileList);
      const count = Math.min(docCount, profileList.length);
      for (let i = 0; i < count; i++) {
        const item = profileList[i];
        const entity = `/${entityRevRouteMap(entityType)}/${item}/hierarchy/nested`;
        debug('entity', entity);
        const result = await graphql(query, { entity });
        await writeTB(entityType, i, 'result', result);
        expect(result.errors).toBeUndefined();
        expect(fixGUIDS(result.data)).toMatchSnapshot();
      }
    });
  });

});
