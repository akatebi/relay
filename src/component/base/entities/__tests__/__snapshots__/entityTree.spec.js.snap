// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entityTree OrganizationHierarchy 1`] = `
Array [
  "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
]
`;

exports[`entityTree OrganizationHierarchy 2`] = `
Object {
  "entityTree": Object {
    "children": Array [
      Object {
        "children": Array [],
        "name": "Corporate Organization",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "Chicago",
              },
              Object {
                "children": Array [],
                "name": "St. Louis",
              },
              Object {
                "children": Array [],
                "name": "Fargo",
              },
              Object {
                "children": Array [],
                "name": "Green Bay",
              },
              Object {
                "children": Array [],
                "name": "Detroit",
              },
            ],
            "name": "Midwest",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "New York",
              },
              Object {
                "children": Array [],
                "name": "Philadelphia",
              },
              Object {
                "children": Array [],
                "name": "Norfolk",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "name": "Providence",
                  },
                  Object {
                    "children": Array [],
                    "name": "Hartford",
                  },
                  Object {
                    "children": Array [],
                    "name": "Portsmouth",
                  },
                  Object {
                    "children": Array [],
                    "name": "Boston",
                  },
                  Object {
                    "children": Array [],
                    "name": "Augusta",
                  },
                ],
                "name": "New England",
              },
              Object {
                "children": Array [],
                "name": "Newark",
              },
            ],
            "name": "Northeast",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "Portland",
              },
              Object {
                "children": Array [],
                "name": "Seattle",
              },
              Object {
                "children": Array [],
                "name": "Boise",
              },
              Object {
                "children": Array [],
                "name": "Billings",
              },
            ],
            "name": "Northwest",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "Nashville",
              },
              Object {
                "children": Array [],
                "name": "Charlotte",
              },
              Object {
                "children": Array [],
                "name": "Cleveland",
              },
              Object {
                "children": Array [],
                "name": "Dallas",
              },
              Object {
                "children": Array [],
                "name": "New Orleans",
              },
              Object {
                "children": Array [],
                "name": "Atlanta",
              },
              Object {
                "children": Array [],
                "name": "Miami",
              },
            ],
            "name": "Southeast",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "Denver",
              },
              Object {
                "children": Array [],
                "name": "Phoenix",
              },
              Object {
                "children": Array [],
                "name": "San Diego",
              },
              Object {
                "children": Array [],
                "name": "Reno",
              },
            ],
            "name": "Southwest",
          },
        ],
        "name": "Enterprise Organization",
      },
    ],
    "name": "All Organizations",
  },
}
`;
